File: SmartReceipts/Model/Price.m
//
//  Price.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 22/04/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "Price.h"

@interface Price ()

@property (nonatomic, strong) NSDecimalNumber *amount;
@property (nonatomic, strong) Currency *currency;

@end

@implementation Price

+ (Price *)priceWithAmount:(NSDecimalNumber *)amount currencyCode:(NSString *)currencyCode {
    Price *price = [[Price alloc] init];
    [price setAmount:amount];
    [price setCurrency:[Currency currencyForCode:currencyCode]];
    return price;
}

+ (Price *)priceWithAmount:(NSDecimalNumber *)amount currency:(Currency *)currency {
    Price *price = [[Price alloc] init];
    [price setAmount:amount];
    [price setCurrency:currency];
    return price;
}

+ (Price *)zeroPriceWithCurrencyCode:(NSString *)currencyCode {
    return [Price priceWithAmount:[NSDecimalNumber zero] currencyCode:currencyCode];
}

- (NSString *)amountAsString {
    return [Price amountAsString:self.amount];
}

+ (NSString *)amountAsString:(NSDecimalNumber *)amount {
    static NSString const *noCurrencyKey = @"noCurrencyFormatter";
    NSNumberFormatter *noCurrencyFormatter = [[NSThread currentThread] threadDictionary][noCurrencyKey];
    if (!noCurrencyFormatter) {
        noCurrencyFormatter = [[NSNumberFormatter alloc] init];
        [noCurrencyFormatter setNumberStyle:NSNumberFormatterDecimalStyle];
        [noCurrencyFormatter setMinimumFractionDigits:2];
        [noCurrencyFormatter setMaximumFractionDigits:2];
        [noCurrencyFormatter setUsesGroupingSeparator:false];
        [[NSThread currentThread] threadDictionary][noCurrencyKey] = noCurrencyFormatter;
    }

    return [[noCurrencyFormatter stringFromNumber:amount] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
}


- (NSString *)mileageRateAmountAsString {
    return [Price mileageRateAmountAsString:self.amount];
}

+ (NSString *)mileageRateAmountAsString:(NSDecimalNumber *)amount {
    static NSString const *noCurrencyRateKey = @"noCurrencyRateFormatter";
    NSNumberFormatter *noCurrencyRateFormatter = [[NSThread currentThread] threadDictionary][noCurrencyRateKey];
    if (!noCurrencyRateFormatter) {
        noCurrencyRateFormatter = [[NSNumberFormatter alloc] init];
        [noCurrencyRateFormatter setNumberStyle:NSNumberFormatterDecimalStyle];
        [noCurrencyRateFormatter setMinimumFractionDigits:2];
        [noCurrencyRateFormatter setMaximumFractionDigits:3];
        [noCurrencyRateFormatter setUsesGroupingSeparator:false];
        [[NSThread currentThread] threadDictionary][noCurrencyRateKey] = noCurrencyRateFormatter;
    }
    
    return [[noCurrencyRateFormatter stringFromNumber:amount] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
}

- (NSString *)currencyFormattedPrice {
    return [self formattedMoneyString:self.amount];
}

- (NSString *)mileageRateCurrencyFormattedPrice {
    NSString *amount = [Price mileageRateAmountAsString:self.amount];
    NSString *currencySymbol = [[NSLocale currentLocale] displayNameForKey:NSLocaleCurrencySymbol value:self.currency.code];
    return [NSString stringWithFormat:@"%@%@", currencySymbol, amount];
}

- (NSString *)formattedMoneyString:(NSDecimalNumber *)moneyAmount {
    return [[Price formatterForCurrencyCode:self.currency.code] stringFromNumber:moneyAmount];
}

+ (NSNumberFormatter *)formatterForCurrencyCode:(NSString *)code {
    NSNumberFormatter *formatter = [[NSThread currentThread] threadDictionary][code];
    if (!formatter) {
        formatter = [[NSNumberFormatter alloc] init];
        [formatter setNumberStyle:NSNumberFormatterCurrencyStyle];
        [formatter setMinimumFractionDigits:2];
        [formatter setMaximumFractionDigits:2];
        [formatter setCurrencyCode:code];
        [[NSThread currentThread] threadDictionary][code] = formatter;
    }

    return formatter;
}

@end

File: SmartReceipts/Models/Price.swift
//
//  Price.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 03/07/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation

@objcMembers
class Price: NSObject {
    private(set) var amount: NSDecimalNumber!
    private(set) var currency: Currency!
    
    init(amount: NSDecimalNumber, currencyCode: String) {
        self.amount = amount
        self.currency = Currency.currency(forCode: currencyCode)
    }
    
    init(amount: NSDecimalNumber, currency: Currency) {
        self.amount = amount
        self.currency = currency
    }
    
    init(currencyCode: String) {
        self.amount = NSDecimalNumber.zero
        self.currency = Currency.currency(forCode: currencyCode)
    }
    
    override init() {
        super.init()
        amount = NSDecimalNumber.zero
        currency = Currency.currency(forCode: WBPreferences.defaultCurrency())
    }
    
    func currencyFormattedPrice() -> String {
        return Price.formatterForCurrency(code: currency.code).string(from: amount)!
    }
    
    func mileageRateCurrencyFormattedPrice() -> String {
        let locale = Locale.current as NSLocale
        let amount = Price.mileageRateStringFrom(amount: self.amount)
        let currencySymbol = locale.displayName(forKey: NSLocale.Key.currencySymbol, value: currency.code) ?? ""
        return "\(currencySymbol)\(amount)"
    }
    
    func amountAsString() -> String {
        return Price.stringFrom(amount: amount)
    }
    
    func mileageRateAmountAsString() -> String {
        return Price.mileageRateStringFrom(amount: amount)
    }
    
    class func stringFrom(amount: NSDecimalNumber) -> String {
        let noCurrencyKey = "noCurrencyFormatter"
        var noCurrencyFormatter = Thread.current.threadDictionary[noCurrencyKey] as? NumberFormatter
        if noCurrencyFormatter == nil {
            noCurrencyFormatter = NumberFormatter()
            noCurrencyFormatter?.numberStyle = .decimal
            noCurrencyFormatter?.maximumFractionDigits = 2
            noCurrencyFormatter?.minimumFractionDigits = 2
            noCurrencyFormatter?.usesGroupingSeparator = false
            Thread.current.threadDictionary[noCurrencyKey] = noCurrencyFormatter
        }
        return (noCurrencyFormatter!.string(from: amount)! as NSString).trimmingCharacters(in: CharacterSet.whitespaces)
    }
    
    class func mileageRateStringFrom(amount: NSDecimalNumber) -> String {
        let noCurrencyRateKey = "noCurrencyRateFormatter"
        var noCurrencyRateFormatter = Thread.current.threadDictionary[noCurrencyRateKey] as? NumberFormatter
        if noCurrencyRateFormatter == nil {
            noCurrencyRateFormatter = NumberFormatter()
            noCurrencyRateFormatter?.numberStyle = .decimal
            noCurrencyRateFormatter?.maximumFractionDigits = 2
            noCurrencyRateFormatter?.minimumFractionDigits = 3
            noCurrencyRateFormatter?.usesGroupingSeparator = false
            Thread.current.threadDictionary[noCurrencyRateKey] = noCurrencyRateFormatter
        }
        return (noCurrencyRateFormatter!.string(from: amount)! as NSString).trimmingCharacters(in: CharacterSet.whitespaces)
    }
    
    class func formatterForCurrency(code: String) -> NumberFormatter {
        var formatter = Thread.current.threadDictionary[code] as? NumberFormatter
        if formatter == nil {
            formatter = NumberFormatter()
            formatter?.numberStyle = .currency
            formatter?.maximumFractionDigits = 2
            formatter?.minimumFractionDigits = 2
            formatter?.currencyCode = code
            Thread.current.threadDictionary[code] = formatter
        }
        return formatter!
    }
}
